packages/block-library/src/embed/edit.js

```diff
@@ -179,6 +152,23 @@ export function getEmbedEditComponent( title, icon, responsive = true ) {
 				);
 			}
 
+			// To render the preview, we must have the attributes determined by the preview.
+			// The higher order `withSelect` component supplies the preview, but cannot
+			// compute the attributes as it does not have access to the internal state of
+			// the edit component. If we wait for `getAttributesFromPreview` to complete
+			// when the component updates with a preview, we will render an `EmbedPreview`
+			// with incorrect props, and this results in a rendering glitch for responsive
+			// embeds. The glitch happens when the content is made responsive by applying
+			// the responsive styles, but the underlying iframe does not get new HTML
+			// for the sandbox to resize, and so it gets the wrong margins and either
+			// gains a lot of unwanted whitespace, or a scrollbar, depending on the size
+			// and shape of the content. So, we get the attributes from the preview here,
+			// and `getAttributesFromPreview` is memoized in `util` so we're not doing
+			// the computation on every render.
```

Comment:
Just to verify my understanding, this comment says: on `componentDidUpdate` the preview is not yet ready and hence `handleIncomingPreview` will attempt to `getAttributesFromPreview` from an unfinished preview, so, _as an extra measure_, we call `getAttributesFromPreview` again at render time, and since it is memoized this is very fast. Right?

Reply:
Kind of. I think maybe I rambled a little bit in that comment and assumed knowledge I shouldn't have...

The problem is that we need to set attributes based on the preview. If we have that preview when the block is constructed, then there are two renders. The first renders the preview. Then when the attributes are worked out and we call `setAttributes`, the second render applies the responsive classes based on the attributes we got from the preview, and the block's internal state. Applying the classes after the initial render causes the whitespace or scrollbar, because the preview needs to be rendered into a container that already has the correct responsive classes.

So, we always get the attributes from the preview in `render`, because the `setAttributes` call won't have completed on the first render.

We memoize `getAttributesFromPreview` because it requires constructing documents, and we don't want those heavyweight things happening on every render.

I know that's also quite rambling... but hopefully clearer this time? :)

Reply:
Much better and clearer, it makes sense now that you explained:

> we always get the attributes from the preview in render, because the setAttributes call won't have completed on the first render.

Reply:
I'll rewrite that comment so it makes sense to people who are not me :) And I'll also address the `getAttributesFromPreview` naming too. Thank you!